src/ert/storage/local_ensemble.py

```diff
@@ -106,6 +107,96 @@ def responses(self) -> npt.NDArray[np.float64]:
         return self._observations_and_responses.iloc[:, 4:].values
 
 
+class RealizationState:
+    def __init__(self) -> None:
+        self._states: Dict[int, Set[Tuple[str, str, bool]]] = {}
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, RealizationState):
+            return self._states == other._states
+
+        return False
+
+    def clear_entry(self, realization: int, key: str) -> None:
+        if realization not in self._states:
+            return
+
+        state_for_real = self._states[realization]
+
+        to_remove = []
+        for tup in state_for_real:
+            _group, _key, _value = tup
+
+            if key in {_group, _key}:
+                to_remove.append(tup)
+
+        for tup in to_remove:
+            state_for_real.remove(tup)
+
+    @staticmethod
+    def from_file(path: Path) -> RealizationState:
+        with open(path, "r") as f:
+            return RealizationState.from_json(json.load(f))
+
+    @staticmethod
+    def from_json(data: Dict[str, List[Tuple[str, str, bool]]]) -> RealizationState:
+        the_state = RealizationState()
+        for realization, entries in data.items():
+            the_state._states[int(realization)] = {
+                (str(group), str(key), bool(value)) for group, key, value in entries
+            }
+
+        return the_state
+
+    def _to_json(self) -> Dict[int, List[Tuple[str, str, bool]]]:
+        return {i: list(states) for i, states in self._states.items()}
+
+    def to_file(self, path: Path) -> None:
+        with open(path, "w+") as f:
+            json.dump(self._to_json(), f)
+
+    def add(self, realization: int, entries: Set[Tuple[str, str, bool]]) -> None:
+        if realization not in self._states:
+            self._states[realization] = set()
+
+        real_state = self._states[realization]
+        assert real_state is not None
+
+        for entry in entries:
+            group, key, value = entry
+            if (group, key, not value) in real_state:
+                real_state.remove((group, key, not value))
+
+        real_state.update(entries)
+
+    def has(self, realization: int, key: str) -> bool:
+        if realization not in self._states:
+            return False
+
+        state = self._states[realization]
+
+        return (
+            (key, key, True) in state
+            or any(_has_it and key == _group for _group, _, _has_it in state)
+            or any(_has_it and key == _key for _, _key, _has_it in state)
+        )
+
+    def has_entry(self, realization: int, key: str) -> bool:
+        if realization not in self._states:
+            return False
+
+        state = self._states[realization]
+
+        result = (
+            (key, key, True) in state
+            or (key, key, False) in state
+            or any(key == _group for _group, _, _ in state)
+            or any(key == _key for _, _key, _ in state)
+        )
+
```

Comment:
What is the difference between these two?

Reply:
has_entry is if a state is registered for the realization, whether it be `True` or `False`, whereas `has` is if there is a state registered and it is also `True`